Actor spawn/despawn node

antonkurkin opened the pull request:
Here is my spawner node proposal
    
When attached to a road point (as a child) it should:
- continuously spawn actors on the lanes that are outgoing from it
- connect internal virtual road lanes to all lanes (in case if some agent moves backward) that are not linked to other lanes. Despawn any actors that are getting to these lanes

Current state:
+ Spawn works and accepts multiple types of actors (actor scenes). done with signals and timers
+ Despawn works. done with despawning lane that frees an actor on asignment
+ Added new icon for the node (supposed to be a factory icon but...)
+ The spawner object does lane search/linking in road point when created/added to the scene tree. also optionally may be connected to segment update signal to reattach on change(just in case but not really used)

Also in this PR: 
- Made procedural demo main scene of the project
- Added anisothropic filtering to road texture(so it would not be blurry)

A fix for https://github.com/TheDuckCow/godot-road-generator/issues/231

antonkurkin commented:
As a side note, I'm thinking whether we should make some sort of policy for who is allowed on a road lane. so aw we wouldn't for example despawn the player if they have to get to the spawner node somehow.
Also could be needed for some routing scenarios - for example lanes/regions where trucks/commercial vehicles are not allowed, or bus-only lanes etc.

antonkurkin commented:
@TheDuckCow would appreciate your thoughts on the direction

TheDuckCow commented:
Thanks for the heads up, let me get back to you after reviewing/thinking it through a bit.

antonkurkin commented:
added some work around for road point transformation signal, not sure how to do properly
made possible to spawn multiple random types of actors

antonkurkin commented:
I tried to add the spawner to some node in intersection demo and it just doesn't work. I think the problem is that it's being assigned before segments are created after the spawner is looking for them. I think I need to do something more

Also tried to move player in procedural demo to another side of the road and it's strange there when moving backwards, it seems like when the point is disconnected, geometry is being deleted while lanes are still connected. need to investigate that further

antonkurkin commented:
so I updated disconnect_roadpoint to actually free and disconnect lanes. not sure if I need to call it directly - or maybe this function should be called from somewhere. anyway now lanes are deleted together with road point and road gemetry.

fixed despawn, now it works as expected.
fixed all the errors (disabled spawning in editor, made RoadAgent a tool - but I doubt that the latter is actually needed?)

also added 0.5s delay to make it work in intersections demo. works now but I still didn't came up with a real solution

antonkurkin commented:
@TheDuckCow I think it's ready for review. Also please let me know where to merge:)

antonkurkin commented:
Oh I set godot editor to remove whitespaces at the end of lines on save (as it was bothering in my changes) and got a lot of unnecessary changes. I can roll it back if you think I should

antonkurkin commented:
moved spawner node to demo, and removed the icon resource. I think I agree that it shouldn't be in the plugin

TheDuckCow commented:
Also FYI I've updated the base to be `dev`, moving forward that'll be the place to put new PRs (not a big deal if you or anyone forgets though, easy to change)

antonkurkin commented:
> Left a few comments for discussion. The most important thing is that it is working well! I have feeling that the solution is a little more on the more complex side, relying on adding several times and extensions of RoadLane types, but that's fine being in a demo scene here.

I actually thought about it a bit more and I think I'll rework it to use 1 timer + 1 despawning lane. otherwise it does seem like an overkill

antonkurkin commented:
ok so I reworked that
Updated spawner to only have one despawner lane and one timer
Created RoadActorManager to reuse despawned actors
Updated spawner and procedural generation demo to use it

also removed class names, rp_max and rename variables according to review comments

antonkurkin commented:
I requested new review, I think it's in good shape but I didn't sleep well this night and I'm not completely sure that I did all that I should have. will re-check tomorrow

antonkurkin commented:
... forgot to add new road actor manager gd script. fixed now

TheDuckCow commented on demo/procedural_generator/road_actor_spawner.gd:
```diff
@@ -0,0 +1,235 @@
+@export var agent_manager_path: NodePath
+## Actor scenes that will be spawned
+## Don't change when the spawner object is attached
+@export var road_actor_scenes: Array[PackedScene]
```
Yup I had the thought of doing this at some point, as well as bring in some other road intersections in his style to pair with a road material that matches. I think it would make more sense to do that update in one singular PR update, but we can keep note that this is well equipped to be added here

TheDuckCow commented on addons/road-generator/nodes/road_segment.gd:
```diff
@@ -577,6 +578,17 @@ func get_lanes() -> Array:
+		if l in ignore_list:
+			return
+		var ln:RoadLane = l.get_node_or_null(l.lane_next)
+		if ln && ln.lane_prior == ln.get_path_to(l):
+			ln.lane_prior = NodePath("")
```
Ah well if there was an actual bug here then by all means I'm for it.

I do agree that doing a better job in the destructors is probably the better option (but then I start swirling in my mind, thinking about how that starts chain reactions in complexity if you remove a RoadContainer with lots of children RoadPoints - you spend all this time cleaning up neighbors that are being deleted anyways). Maybe we could conditionally do queue_free() on sibling/connected nodes only if their parents aren't already marked for deletion or something like that.

This would likely also need to pair with changing the somewhat annoying "auto fix" code that exists right now, making things a bit complicated. I have a separate ticket here talking about some of that: https://github.com/TheDuckCow/godot-road-generator/issues/238

Might be good to push more of the discussion on cleanup over there, so that whatever we decide can be applied more universally.

antonkurkin commented:
while playing around found that I should have unregistered actor from lane when hiding it for reuse

TheDuckCow reviewed (approved):
Overall I think things are in a good state - performance is good, works as expected, and the code is generally nicely cleaner. Just a couple minor suggestions here but once you make those little changes (and not sure if you have any other cleanup planned).

Let me know if there's anything more specific you want to discuss, but excited to have this merged in

antonkurkin commented on demo/procedural_generator/road_actor_manager.gd:
I was thinking of keeping kind of internal consistency in our terms. like it is a structure to keep track/create/remove actors (vehicle objects)
Traffic manager is a good name, but maybe it would be a better name to an object that manages traffic (like a process - actual movement of actors) on intersections or something like that?

antonkurkin commented on demo/procedural_generator/road_actor_spawner.gd:
```diff
@@ -63,173 +109,168 @@ func _exit_tree():
+		if ! new_actor_manager:
+			push_error("Actor manager path ", new_path, " is incorrect")
+		elif ! new_actor_manager.has_method("add_actor") || ! new_actor_manager.has_method("remove_actor"):
+			push_error("Actor manager at ", new_path, " shall have add_actor and remove_actor methods")
```
yeah, shall is kind of widely used in my field (C standard) but yeah I'll replace them for ease of reading

antonkurkin commented:
@TheDuckCow updated error messages
I don't have anything specific to discuss (still didn't get to https://github.com/TheDuckCow/godot-road-generator/issues/238)
And for the core of spawner node I don't have any improvements ideas. I'm trying to make actors aware of other actors and avoid collisions, that will lead to proposals for changes in road agent/road lane interfaces. but it will be in another PR...

So I think it's ready to be merged:)

antonkurkin commented:
ah only one thing is maybe about tests but as both actor manager and actor spawner are basically an example code, I'm not sure we want that?
